### PR TITLE
[MM-16697] Rework validation of team and channel membership for webhook updates

### PR DESCRIPTION
**Pull request by DSchalla**
#### Summary
This PR reworks the validation of the channel permissions before a webhook update is permitted.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-16697

### CODE REVIEW

File: api4/webhook_test.go [hunk 1]
```diff
@@ -74,6 +74,51 @@ func TestCreateIncomingWebhook(t *testing.T) {
...
+func TestCreateIncomingWebhook_BypassTeamPermissions(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableIncomingWebhooks = true })
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnablePostUsernameOverride = true })
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnablePostIconOverride = true })
+
+	defaultRolePermissions := th.SaveDefaultRolePermissions()
+	defer func() {
```
Comment (mgdelacroix) [inline api4/webhook_test.go]: I think directly doing

```go
defer th.RestoreDefaultRolePermissions(defaultRolePermissions)
```

should work fine

File: api4/webhook_test.go [hunk 2]
```diff
@@ -74,6 +74,51 @@ func TestCreateIncomingWebhook(t *testing.T) {
...
+	th.RemovePermissionFromRole(model.PERMISSION_MANAGE_INCOMING_WEBHOOKS.Id, model.SYSTEM_USER_ROLE_ID)
+	th.AddPermissionToRole(model.PERMISSION_MANAGE_INCOMING_WEBHOOKS.Id, model.TEAM_ADMIN_ROLE_ID)
+	th.AddPermissionToRole(model.PERMISSION_MANAGE_INCOMING_WEBHOOKS.Id, model.TEAM_USER_ROLE_ID)
+
+	hook := &model.IncomingWebhook{ChannelId: th.BasicChannel.Id}
+
+	rhook, resp := th.Client.CreateIncomingWebhook(hook)
+	CheckNoError(t, resp)
+
+	if rhook.ChannelId != hook.ChannelId {
```
Comment (mgdelacroix) [inline api4/webhook_test.go]: Just curious, have you found a case where the `ChannelId`, `UserId` or `TeamId` checks fail?

File: api4/webhook_test.go [hunk 3]
```diff
@@ -74,6 +74,51 @@ func TestCreateIncomingWebhook(t *testing.T) {
...
+		t.Fatal("channel ids didn't match")
+	}
+
+	if rhook.UserId != th.BasicUser.Id {
```
Comment (mgdelacroix) [inline api4/webhook_test.go]: nit: you can use `require.Equal` to improve readability here

File: api4/webhook_test.go (hunk 2 above)
Comment (DSchalla) [reply api4/webhook_test.go]: Not really. I copied it mostly to mirror the regular test case above.

File: api4/webhook_test.go [hunk 4]
```diff
@@ -639,6 +674,50 @@ func TestUpdateIncomingHook(t *testing.T) {
...
+func TestUpdateIncomingWebhook_BypassTeamPermissions(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableIncomingWebhooks = true })
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnablePostUsernameOverride = true })
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnablePostIconOverride = true })
+
+	defaultRolePermissions := th.SaveDefaultRolePermissions()
+	defer func() {
```
Comment (mgdelacroix) [inline api4/webhook_test.go]: Same case for this `defer`

File: api4/webhook_test.go [hunk 5]
```diff
@@ -639,6 +674,50 @@ func TestUpdateIncomingHook(t *testing.T) {
...
+	th.RemovePermissionFromRole(model.PERMISSION_MANAGE_INCOMING_WEBHOOKS.Id, model.SYSTEM_USER_ROLE_ID)
+	th.AddPermissionToRole(model.PERMISSION_MANAGE_INCOMING_WEBHOOKS.Id, model.TEAM_ADMIN_ROLE_ID)
+	th.AddPermissionToRole(model.PERMISSION_MANAGE_INCOMING_WEBHOOKS.Id, model.TEAM_USER_ROLE_ID)
+
+	hook := &model.IncomingWebhook{ChannelId: th.BasicChannel.Id}
+
+	rhook, resp := th.Client.CreateIncomingWebhook(hook)
+	CheckNoError(t, resp)
+
+	if rhook.ChannelId != hook.ChannelId {
```
Comment (mgdelacroix) [inline api4/webhook_test.go]: Same for this checks to be `require`s

File: api4/webhook_test.go [hunk 6]
```diff
@@ -834,6 +913,45 @@ func TestUpdateOutgoingHook(t *testing.T) {
...
 }
 
+func TestUpdateOutgoingWebhook_BypassTeamPermissions(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableOutgoingWebhooks = true })
+
+	defaultRolePermissions := th.SaveDefaultRolePermissions()
+	defer func() {
```
Comment (mgdelacroix) [inline api4/webhook_test.go]: Same for this `defer`

File: api4/webhook_test.go [hunk 7]
```diff
@@ -834,6 +913,45 @@ func TestUpdateOutgoingHook(t *testing.T) {
...
+	th.AddPermissionToRole(model.PERMISSION_MANAGE_OUTGOING_WEBHOOKS.Id, model.TEAM_ADMIN_ROLE_ID)
+	th.AddPermissionToRole(model.PERMISSION_MANAGE_OUTGOING_WEBHOOKS.Id, model.TEAM_USER_ROLE_ID)
+
+	hook := &model.OutgoingWebhook{ChannelId: th.BasicChannel.Id, TeamId: th.BasicChannel.TeamId,
+		CallbackURLs: []string{"http://nowhere.com"}, TriggerWords: []string{"rats2"}}
+
+	rhook, resp := th.Client.CreateOutgoingWebhook(hook)
+	CheckNoError(t, resp)
+
+	if rhook.ChannelId != hook.ChannelId {
```
Comment (mgdelacroix) [inline api4/webhook_test.go]: Same for this checks to be `require`s

File: api4/webhook.go [hunk 8]
```diff
@@ -96,20 +96,20 @@ func updateIncomingHook(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	channel, err := c.App.GetChannel(updatedHook.ChannelId)
+	if err != nil {
+		c.Err = err
 		return
 	}
 
-	if c.App.Session.UserId != oldHook.UserId && !c.App.SessionHasPermissionToTeam(c.App.Session, updatedHook.TeamId, model.PERMISSION_MANAGE_OTHERS_INCOMING_WEBHOOKS) {
-		c.LogAudit("fail - inappropriate permissions")
-		c.SetPermissionError(model.PERMISSION_MANAGE_OTHERS_INCOMING_WEBHOOKS)
+	if !c.App.SessionHasPermissionToTeam(c.App.Session, channel.TeamId, model.PERMISSION_MANAGE_INCOMING_WEBHOOKS) {
```
Comment (crspeller) [inline api4/webhook.go]: Should we check `channel.TeamId == updatedHook.TeamId`?

Comment (grundleborg) [reply api4/webhook.go]: Yes :)

Comment (DSchalla) [reply api4/webhook.go]: Yup sounds good to me.

**Pull request merged**